File: app/src/main/java/crabster/rudakov/sberschoollesson30hwk/domain/interactors/DictionaryInteractorImpl.java
package crabster.rudakov.sberschoollesson30hwk.domain.interactors;

import androidx.annotation.*;

import java.util.*;

import crabster.rudakov.sberschoollesson30hwk.domain.converter.*;
import crabster.rudakov.sberschoollesson30hwk.domain.models.*;
import crabster.rudakov.sberschoollesson30hwk.domain.repositories.*;
import io.reactivex.*;

public class DictionaryInteractorImpl implements IDictionaryInteractor {

    private final IDictionaryRepository repository;
    private final IDictionaryConverter converter;

    public DictionaryInteractorImpl(@NonNull IDictionaryRepository repository,
                                    @NonNull IDictionaryConverter converter) {
        this.repository = repository;
        this.converter = converter;
    }

    @Override
    public Completable add(@NonNull DictionaryItem dictionaryItem) {
        return repository.add(converter.convert(dictionaryItem));
    }

    @Override
    public Single<List<DictionaryItem>> getList() {
        return repository.getList()
                .map(converter::reverseList);
    }

    @Override
    public Completable delete(long id) {
        return repository.delete(id);
    }

}

File: app/src/main/java/crabster/rudakov/sberschoollesson30hwk/domain/interactors/DictionaryInteractorImpl.kt
package crabster.rudakov.sberschoollesson30hwk.domain.interactors

import crabster.rudakov.sberschoollesson30hwk.domain.converter.IDictionaryConverter
import crabster.rudakov.sberschoollesson30hwk.domain.models.DictionaryItem
import crabster.rudakov.sberschoollesson30hwk.domain.repositories.IDictionaryRepository
import crabster.rudakov.sberschoollesson30hwk.domain.repositories.models.DictionaryItemModel
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Single

class DictionaryInteractorImpl(
    private val repository: IDictionaryRepository,
    private val converter: IDictionaryConverter
) : IDictionaryInteractor {

    override fun add(dictionaryItem: DictionaryItem): Completable {
        return repository.add(converter.convert(dictionaryItem))
    }

    override fun getList(): Single<List<DictionaryItemModel>> {
        return repository.getList()
    }

    override fun delete(id: Long): Completable {
        return repository.delete(id)
    }

}